File: chap18_gastrointestinal/exercises/ex_coupled_FHN_chain.py

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt
from scipy.signal import find_peaks

def deriv(t, y):
    """ODEs for FHN."""
    V = y[0:n]
    w = y[n:2*n]
    alpha = 0.4
    eps = np.linspace(0.02,0.08,num=n)
    Vdot = np.empty(n)
    
    wdot = V - alpha
    Vdot = (1/eps)*( V*(1-V)*(V-alpha) - w)
    for i in range(1,n-2):
        Vdot[i] = Vdot[i] + G*(V[i+1]-V[i]) + G*(V[i-1]-V[i])
    Vdot[0] = Vdot[0]   + G*(V[1]-V[0]) 
    Vdot[n-1] = Vdot[n-1]  + G*(V[n-2]-V[n-1])
    
    return np.concatenate((Vdot,wdot))


n = 200   # number of oscillators in chain

# Output times and range
t0=0
tf=60
t = np.arange(t0,tf,0.05)
# Initial conditions for all oscillators: V = 0.1; w = 0.
y0 = 0.1*np.ones(2*n)
y0[n:2*n] = 0

# First solve without coupling
G=0
soln = solve_ivp(deriv, (t0, tf), y0, method='Radau',t_eval=t)

# Find the peaks of each oscillator, and thus calculate the frequency
freq = np.empty(n)
for i in range(0, n):
    peaks,_ = find_peaks(soln.y[i])
    peak_times = t[peaks]
    freq[i] = 1/((peak_times[-1] - peak_times[-10])/10)

# Plot solution
plt.plot(freq)

# Now solve with coupling
G=50
soln = solve_ivp(deriv, (t0, tf), y0, method='Radau',t_eval=t)

# Find the peaks of each oscillator, and thus calculate the frequency
freq = np.empty(n)
for i in range(0, n):
    peaks,_ = find_peaks(soln.y[i])
    peak_times = t[peaks]
    freq[i] = 1/((peak_times[-1] - peak_times[-10])/10)

# Plot solution
plt.plot(freq)

File: chap18_gastrointestinal/exercises/ex_coupled_FHN_chain.m
function main

close all
clear all
clc

n = 200 ; % number of coupled oscillators

% First plot the uncoupled case, to see the natural frequencies
G=0;
tspan = linspace(0,60,1200);
x0(1:n) = 0.1; %linspace(0,0.8,n);
x0(n+1:2*n) = 0;
[t,sol] = ode15s(@(t,x) FHN(t,x,n,G),tspan,x0);
%plot(t,sol(:,2),t,sol(:,100),t,sol(:,200))

% Find the frequency of each oscillator
for i=1:n
    [~,idx] = findpeaks(sol(:,i));
    period(i) = (t(idx(end)) - t(idx(end-10)))/10; % average over 10 cycles to get smoother result
    frequency(i) = 1/period(i);
end
plot(frequency)
hold on


% Now plot the coupled case
G=50;
[t,sol] = ode15s(@(t,x) FHN(t,x,n,G),tspan,x0);

% Find the frequency of each oscillator
for i=1:n
    [~,idx] = findpeaks(sol(:,i));
    period(i) = (t(idx(end)) - t(idx(end-10)))/10; % average over 10 cycles to get smoother result
    frequency(i) = 1/period(i);
end
plot(frequency)

end

% -------------------------------------------------------------------------------
%% FHN equations
function dxdt = FHN(t,x,n,G)

v = x(1:n);
w = x(n+1:2*n);

alpha = 0.4;
eps = linspace(0.02,0.08,n);

dxdt(n+1:2*n) = v - alpha';  % no coupling in w
for i=2:n-1
    dxdt(i) = (1/eps(i))*(v(i).*(v(i)-1).*(alpha - v(i)) - w(i)) + G*(v(i+1)-v(i)) + G*(v(i-1)-v(i));
end
dxdt(1) = (1/eps(i))*(v(1).*(v(1)-1).*(alpha - v(1)) - w(1)) + G*(v(2)-v(1));
dxdt(n) = (1/eps(i))*(v(n).*(v(n)-1).*(alpha - v(n)) - w(n)) + G*(v(n-1)-v(n));

dxdt = dxdt';
end